lib/programmemory.cpp

```diff
@@ -173,14 +173,27 @@ void ProgramMemory::erase_if(const std::function<bool(const ExprIdToken&)>& pred
     if (mValues->empty())
         return;
 
-    // TODO: how to delay until we actuallly modify?
-    copyOnWrite();
+    if (mValues.use_count() == 1)
+    {
+        // we are the only user so can use it directly
+        for (auto it = mValues->begin(); it != mValues->end();) {
+            if (pred(it->first))
+                it = mValues->erase(it);
+            else
+                ++it;
+        }
+    }
```

Comment:
I dont think this extra branch is needed. Especially if you can move the `copyOnWrite` out of the loop.